frontend/components/markup-modal.js

```diff
@@ -2,86 +2,163 @@ import React from 'react'
 import PropTypes from 'prop-types'
 import Modal from 'react-modal'
 import ClickToSelect from '@mapbox/react-click-to-select'
-import resolveBadgeUrl from '../lib/badge-url'
+import { badgeUrlFromPath, badgeUrlFromPattern } from '../../lib/make-badge-url'
 import generateAllMarkup from '../lib/generate-image-markup'
 import { advertisedStyles } from '../../supported-features.json'
 
+const nonBreakingSpace = '\u00a0'
+
 export default class MarkupModal extends React.Component {
   static propTypes = {
-    example: PropTypes.shape({
-      title: PropTypes.string.isRequired,
-      exampleUrl: PropTypes.string,
-      previewUrl: PropTypes.string,
-      urlPattern: PropTypes.string,
-      documentation: PropTypes.string,
-      link: PropTypes.string,
-    }),
+    example: PropTypes.object,
```

Comment:
Are we losing some useful validation here?

Reply:
Part of what makes prop types useful is that they gives a single point of reference on the props schema. Since the service definition data is passed straight through to the components – this is one of the `example` keys from it – the service definition schema can be used as a reference.

It seems like a lot of work to maintain this schema in two places, in two incompatible formats.

Prop types also so not prevent components from incorrectly referencing props, or referencing props that don't exist. Related to that… in #1272 @Daniel15 suggested adding typing to the repo. This changeset took many hours to produce, and most of the reason is that the types were changing drastically, from the top all the way down. Type checking seems pretty essential for frontends that can be refactored with ease. I suspect adding TypeScript might make the frontend easier for the average potential frontend contributor, despite the added learning curve. 

I continue to think this is not as pressing an issue for the server code and definitely think it will deter some contributors. JavaScript, even with ES6, is a least-common-denominator kind of language. It's pretty easy for anyone to read, and even easier to write.

I'm wanting to make significant improvements to the frontend, and I'm thinking we may need to recruit some frontend contributors in a targeted way in order to do that. It's on my mind to create a front end dev experience that's as good as the back end dev experience. Of course, we also need to spend some time on the design (#1899) so that can happen in an organized way.

But that's all an aside. TypeScript is hardly pressing.

tldr; I think the validation is less important than the documentation. Maybe adding a link to the schema documentation would be sufficient.

Reply:
There's quite a lot of stuff there, but I think the key point is

> Prop types also so not prevent components from incorrectly referencing props, or referencing props that don't exist

:+1:

Reply:
> Part of what makes prop types useful is that they gives a single point of reference on the props schema. Since the service definition data is passed straight through to the components – this is one of the example keys from it – the service definition schema can be used as a reference.

You can reuse prop types; you don't have to redefine them everywhere. You could make a module that exports the prop types
```
module.exports = {
  Example: PropTypes.shape({
    title: PropTypes.string.isRequired,
    exampleUrl: PropTypes.string,
    previewUrl: PropTypes.string,
    urlPattern: PropTypes.string,
    documentation: PropTypes.string,
    link: PropTypes.string,
  }),
};
```

Then you could do `example: MyShapes.Example` or whatever here.

That's pretty similar to how you'd do this with Flow or TypeScript - Define the type in some reusable place, and reuse it everywhere.

Reply:
Hi! Right, thanks for that suggestion. I have done that in some places. The issue here is different: [the schema is defined using Joi](https://github.com/badges/shields/blob/master/services/service-definitions.js).

That's the validation framework we're using for the JSON responses we get back, which we adopted for runtime use without a ton of consideration (beyond it being integrated into our API testing framework).

That said, last I looked, I did not find a library that could do runtime validation of plain JS objects using TypeScript types.

There do seem to be some attempts to use Joi with React. I didn't dig too far into [the most promising example](https://www.npmjs.com/package/react-joi-validation), though I also wouldn't want to ship Joi to the frontend.

I wouldn't mind using it in dev, though it seems like it would take some effort to configure webpack to accomplish that.